spring-integration-kafka/src/main/java/org/springframework/integration/kafka/inbound/KafkaInboundEndpoint.java

```diff
@@ -64,16 +65,24 @@ default void doWithRetry(RetryTemplate template, RecoveryCallback<?> callback, O
 
 		try {
 			template.execute(context -> {
-				context.setAttribute(CONTEXT_RECORD, data);
-				context.setAttribute(CONTEXT_ACKNOWLEDGMENT, acknowledgment);
-				context.setAttribute(CONTEXT_CONSUMER, consumer);
+				if (context.getRetryCount() == 0) {
+					context.setAttribute(CONTEXT_RECORD, data);
+					context.setAttribute(CONTEXT_ACKNOWLEDGMENT, acknowledgment);
+					context.setAttribute(CONTEXT_CONSUMER, consumer);
+					getAttributesHolder().set(context);
+				}
 				runnable.run();
 				return null;
 			}, callback);
 		}
 		catch (Exception ex) {
 			throw new KafkaException("Failed to execute runnable", ex);
 		}
+		finally {
+			getAttributesHolder().remove();
```

Comment:
I wonder if instead of this method implementation requirement we can move that `ATTRIBUTES_HOLDER` into this interface.
All the impls of this one works in their onw thread and there really not possible to have several of them in the same thread.
So, I don't see any problems to have that constant on the interface.

Reply:
Yes, that looks more concise.